src/database_migration.py

```diff
@@ -0,0 +1,142 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the DatabaseMigration class to manage database migrations."""
+
+import logging
+from typing import Literal, cast
+
+import ops
+from ops.pebble import ExecError
+
+from constants import FLASK_STATE_DIR
+from exceptions import CharmConfigInvalidError
+
+logger = logging.getLogger(__name__)
+
+
+DatabaseMigrationStatus = Literal["PENDING", "COMPLETED", "FAILED"]
+
+
+class DatabaseMigration:
+    """The DatabaseMigration class that manages database migrations.
+
+    Attrs:
+        script: the database migration script.
+        COMPLETED: one of the possible database migration status.
+        FAILED: one of the possible database migration status.
+        PENDING: one of the possible database migration status.
+    """
+
+    _STATUS_FILE = FLASK_STATE_DIR / "database-migration-status"
+    _COMPLETED_SCRIPT_FILE = FLASK_STATE_DIR / "completed-database-migration"
+
+    COMPLETED: DatabaseMigrationStatus = "COMPLETED"
+    FAILED: DatabaseMigrationStatus = "FAILED"
+    PENDING: DatabaseMigrationStatus = "PENDING"
+
+    def __init__(self, flask_container: ops.Container, database_migration_script: str | None):
```

Comment:
Under what conditions would `database_migration_script` be `None` please? I guess if the user doesn't have a migration script defined?

Reply:
Yes, if user doesn't config the `database_migration_script` charm configuration.